src/vos/vos_common.c

```diff
@@ -151,16 +151,15 @@ vos_tx_publish(struct dtx_handle *dth, int err)
 		if (rc == 0 || err != 0)
 			rc = vos_publish_blocks(cont, &dru->dru_nvme,
```

Comment:
so vos_publish_blocks() won't be called once some vos_publish_scm() failed, then nobody will publish or cancel the remaining reserved blocks.

I think this function could be simplified further as:
static int vos_tx_publish(struct dtx_handle *dth, bool publish)
{
    for (....) {
        if (dru->dru_scm != NULL) {
            rc = vos_publish_scm(publish);
            D_FREE(dru->dru_scm);
            dru->dru_scm = NULL;
        }
        if (rc && publish)
            return rc;
        if (!d_list_empty(dru->dru_nvme)
            rc = vos_publish_blocks(publish);
        ....
}

So that we can call vos_tx_publish(true) to do publish or vos_tx_publish(false) to do cancel at any point.

Reply:
reworked it again